src/python/pants/goal/run_tracker.py

```diff
@@ -414,21 +414,93 @@ def shutdown_worker_pool(self):
     """
     SubprocPool.shutdown(self._aborted)
 
-  def report_target_info(self, scope, target, key, val):
+  @classmethod
+  def _create_dict_with_nested_keys_and_val(cls, keys, value):
+    """Recursively constructs a nested dictionary with the keys pointing to the value.
+
+    For example:
+    Given the list of keys ['a', 'b', 'c', 'd'] and a primitive
+    value 'hello world', the method will produce the nested dictionary
+    {'a': {'b': {'c': {'d': 'hello world'}}}}. The number of keys in the list
+    defines the depth of the nested dict. If the list of keys is ['a'] and
+    the value is 'hello world', then the result would be {'a': 'hello world'}.
+
+    :param list of string keys: A list of keys to be nested as a dictionary.
+    :param primitive value: The value of the information being stored.
+    :return: dict of nested keys leading to the value.
+    """
+    current_index = len(keys) - 1
+
+    if len(keys) > 1:
+      new_val = {keys[current_index]: value}
+      new_keys = keys[:current_index]
+      return cls._create_dict_with_nested_keys_and_val(new_keys, new_val)
+    elif len(keys) == 1:
+      return {keys[current_index]: value}
```

Comment:
and `keys[0]` here, and thus get rid of `current_index` entirely!

Sometimes Python can be pretty neat!